python/sbp/client/framer.py

```diff
@@ -100,52 +146,121 @@ def _readall(self, size):
         size : int
           Number of bytes to read.
         """
-        data = b""
-        while len(data) < size:
-            d = self._read(size - len(data))
-            if self._broken:
-                raise StopIteration
-            if not d:
-                # NOTE (Buro/jgross): Force a yield here to another thread. In
-                # case the stream fails midstream, the spinning here causes
-                # the UI thread to lock up without yielding.
-                time.sleep(0)
-                continue
-            data += d
-        return data
+        if self._rx_buf_unprocessed < size:
+            if self._greedy:
+                data = self._read(len(self._rx_buf) - self._rx_buf_unprocessed)
+            else:
+                data = self._read(size)
+
+            l = len(data)
+
+            while l < size:
+                d = self._read(size - len(data))
+                if not d or self._broken:
+                    if allow_empty:
+                        return None
+                    raise StopIteration
+                data += d
+
+            self._rx_buf = self._shift(self._rx_buf, -self._rx_buf_idx)
+            self._rx_buf_idx = 0
+            l = len(data)
+            self._rx_buf[self._rx_buf_unprocessed:self._rx_buf_unprocessed + l] = bytearray(data)
+            self._rx_buf_unprocessed += l
+
+        mv = memoryview(self._rx_buf)[self._rx_buf_idx:self._rx_buf_idx + size]
+        self._rx_buf_unprocessed -= size
+        self._rx_buf_idx += size
+
+        return mv
+
+    def _collect_into(self, size, allow_empty=False):
+        """
+        Read until all bytes are collected.
+
+        Parameters
+        ----------
+        size : int
+          Number of bytes to read.
+        """
+        if self._rx_buf_unprocessed < size:
+            if self._greedy:
+                to_read = len(self._rx_buf) - self._rx_buf_unprocessed
+            else:
+                to_read = size - self._rx_buf_unprocessed
+
+            # Bytes are read into the tail of the buffer. If the index is too
+            # far to fit the new data, buffer needs to be rolled.
+            free_tail = len(self._rx_buf) - self._rx_buf_idx
+            if free_tail < to_read:
+                self._rx_buf = self._shift(self._rx_buf, -self._rx_buf_idx)
```

Comment:
We should try to eliminate this since it is effectively a copy.  I think we could use a "double buffer" idea to implement this, that is:

  1. allocate two buffers for all reads (say 4k each), buffer A and buffer B
  2. buffer A is initially assigned as the "read buffer", buffer B is the "leftover" buffer
  3. "read buffer" is passed to network/file layer that supports `readinto`
  4. read buffer is returned filled with data
  5. SBP processing is done, some amount of data is potentially left over, this becomes the "left over buffer", so "read buffer" and "leftover buffer" swap
  6. It's an error if 4k of "left-over" is accumulated and the left-over buffer is flushed

Reply:
I can try that. Although I tried with big enough buffer size to eliminate rolling altogether and didn't see improvement.